src/Composer/Repository/ArrayRepository.php

```diff
@@ -121,15 +122,13 @@ public function search($query, $mode = 0, $type = null)
      */
     public function hasPackage(PackageInterface $package)
     {
-        $packageId = $package->getUniqueName();
-
-        foreach ($this->getPackages() as $repoPackage) {
-            if ($packageId === $repoPackage->getUniqueName()) {
-                return true;
+        if (empty($this->packageMap)) {
```

Comment:
1. I think it's better to compare with null, because package map potentially could be empty array.
2. We need to make $this->packageMap an empty array in case if it wasn't initialized
```suggestion
        if ($this->packageMap === null) {
            $this->packageMap = [];
```

Reply:
I implemented 1 & 2. 

I had a look at implementing 3 but the problem with that is that this changes the public API of getPackages(). With the new implementation getPackages() would return an array with string indexes vs int indexes previously. To implement your suggestion I made these changes:

```
diff --git a/src/Composer/Repository/ArrayRepository.php b/src/Composer/Repository/ArrayRepository.php
index 3d9e22d1..8331d011 100644
--- a/src/Composer/Repository/ArrayRepository.php
+++ b/src/Composer/Repository/ArrayRepository.php
@@ -28,7 +28,6 @@ class ArrayRepository extends BaseRepository
 {
     /** @var PackageInterface[] */
     protected $packages;
-    protected $packageMap;
 
     public function __construct(array $packages = array())
     {
@@ -122,14 +121,11 @@ public function search($query, $mode = 0, $type = null)
      */
     public function hasPackage(PackageInterface $package)
     {
-        if ($this->packageMap === null) {
-            $this->packageMap = array();
-            foreach ($this->getPackages() as $repoPackage) {
-                $this->packageMap[$repoPackage->getUniqueName()] = $repoPackage;
-            }
+        if (null === $this->packages) {
+            $this->initialize();
         }
 
-        return isset($this->packageMap[$package->getUniqueName()]);
+        return isset($this->packages[$package->getUniqueName()]);
     }
 
     /**
@@ -143,7 +139,7 @@ public function addPackage(PackageInterface $package)
             $this->initialize();
         }
         $package->setRepository($this);
-        $this->packages[] = $package;
+        $this->packages[$package->getUniqueName()] = $package;
 
         if ($package instanceof AliasPackage) {
             $aliasedPackage = $package->getAliasOf();
@@ -165,14 +161,8 @@ protected function createAliasPackage(PackageInterface $package, $alias, $pretty
      */
     public function removePackage(PackageInterface $package)
     {
-        $packageId = $package->getUniqueName();
-
-        foreach ($this->getPackages() as $key => $repoPackage) {
-            if ($packageId === $repoPackage->getUniqueName()) {
-                array_splice($this->packages, $key, 1);
-
-                return;
-            }
+        if (isset($this->packages[$package->getUniqueName()])) {
+            unset($this->packages[$package->getUniqueName()]);
         }
     }
 
diff --git a/tests/Composer/Test/Repository/ArrayRepositoryTest.php b/tests/Composer/Test/Repository/ArrayRepositoryTest.php
index 1c7972c9..f328ca31 100644
--- a/tests/Composer/Test/Repository/ArrayRepositoryTest.php
+++ b/tests/Composer/Test/Repository/ArrayRepositoryTest.php
@@ -39,7 +39,7 @@ public function testRemovePackage()
         $repo->removePackage($this->getPackage('foo', '1'));
 
         $this->assertCount(1, $repo);
-        $this->assertEquals(array($package), $repo->getPackages());
+        $this->assertEquals(array($package->getUniqueName() => $package), $repo->getPackages());
     }
 
     public function testHasPackage()

```

this resulted in these tests failing:

```
vendor/bin/phpunit tests                                                       
PHPUnit 4.8.36 by Sebastian Bergmann and contributors.

...............S.............................................   61 / 1186 (  5%)
.............................................................  122 / 1186 ( 10%)
.....................................F.......................  183 / 1186 ( 15%)
..................................F..........................  244 / 1186 ( 20%)
.............................................................  305 / 1186 ( 25%)
.............................................................  366 / 1186 ( 30%)
.............................................................  427 / 1186 ( 36%)
.............................................................  488 / 1186 ( 41%)
.............................................................  549 / 1186 ( 46%)
..............................S..............................  610 / 1186 ( 51%)
.............................................................  671 / 1186 ( 56%)
.............................................................  732 / 1186 ( 61%)
.............................................................  793 / 1186 ( 66%)
..........................F..........E.....E.....EEE.........  854 / 1186 ( 72%)
...SS........................................................  915 / 1186 ( 77%)
.............................................................  976 / 1186 ( 82%)
............................................................. 1037 / 1186 ( 87%)
............................................................. 1098 / 1186 ( 92%)
............................................................. 1159 / 1186 ( 97%)
...................S.......

Time: 8.45 seconds, Memory: 40.00MB

There were 5 errors:

1) Composer\Test\Repository\CompositeRepositoryTest::testGetPackages
Undefined offset: 0

/data/src/composer/tests/Composer/Test/Repository/CompositeRepositoryTest.php:94

2) Composer\Test\Repository\FilesystemRepositoryTest::testRepositoryRead
Undefined offset: 0

/data/src/composer/tests/Composer/Test/Repository/FilesystemRepositoryTest.php:40

3) Composer\Test\Repository\PathRepositoryTest::testLoadPackageFromFileSystemWithoutVersion
Undefined offset: 0

/data/src/composer/tests/Composer/Test/Repository/PathRepositoryTest.php:68

4) Composer\Test\Repository\PathRepositoryTest::testLoadPackageFromFileSystemWithWildcard
Undefined offset: 0

/data/src/composer/tests/Composer/Test/Repository/PathRepositoryTest.php:90

5) Composer\Test\Repository\PathRepositoryTest::testUrlRemainsRelative
Undefined offset: 0

/data/src/composer/tests/Composer/Test/Repository/PathRepositoryTest.php:123

--

There were 3 failures:

1) Composer\Test\DependencyResolver\DefaultPolicyTest::testSelectLocalReposFirst
Failed asserting that Array &0 (
    0 => 2
) is identical to Array &0 (
    0 => -1
).

/data/src/composer/tests/Composer/Test/DependencyResolver/DefaultPolicyTest.php:203

2) Composer\Test\DependencyResolver\SolverTest::testInstallOneOfTwoAlternatives
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
             'archiveExcludes' => Array ()
-            'id' => -1
+            'id' => 1
             'name' => 'a'
             'prettyName' => 'A'
             'repository' => Composer\Repository\ArrayRepository Object (...)
             'transportOptions' => Array ()
         )
     )
 )

/data/src/composer/tests/Composer/Test/DependencyResolver/SolverTest.php:939
/data/src/composer/tests/Composer/Test/DependencyResolver/SolverTest.php:432

3) Composer\Test\Repository\ComposerRepositoryTest::testLoadData with data set #1 (array(array('bar/foo', '3.14'), array('bar/foo', '3.145')), array(array(array(array('bar/foo', '3.14'), array('bar/foo', '3.145')))))
Failed asserting that actual size 1 matches expected size 2.

/data/src/composer/tests/Composer/Test/Repository/ComposerRepositoryTest.php:62

FAILURES!
Tests: 1186, Assertions: 2395, Errors: 5, Failures: 3, Skipped: 5.
```

Most of them simply seem to fail because they access the packages array with the previously used int id. I'm not too familiar with composer. I'm afraid that this change would impact things beyond tests. Maybe someone with solid knowledge of composer could chime in here. I would be happy to see this through in case you want to go down the uniqueName index route. We could also keep the public API as is and use the implementation with the dedicated map. This would definitely pose the lowest risk.